lib/internal/modules/esm/loader.js

```diff
@@ -202,33 +210,66 @@ class ESMLoader {
       const { ModuleWrap, callbackMap } = internalBinding('module_wrap');
       const module = new ModuleWrap(url, undefined, source, 0, 0);
       callbackMap.set(module, {
-        importModuleDynamically: (specifier, { url }) => {
-          return this.import(specifier, url);
+        importModuleDynamically: (specifier, { url }, import_assertions) => {
+          return this.import(specifier, url, import_assertions);
         }
       });
 
       return module;
     };
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
+    finalFormatCache.set(job, 'module');
     const { module } = await job.run();
 
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async getModuleJob(specifier, parentURL) {
+  async getModuleJob(specifier, parentURL, import_assertions) {
+    if (!ArrayPrototypeIncludes(supportedTypes, import_assertions.type)) {
+      throw new ERR_INVALID_IMPORT_ASSERTION('type', import_assertions.type);
+    }
+
     const { format, url } = await this.resolve(specifier, parentURL);
     let job = this.moduleMap.get(url);
     // CommonJS will set functions for lazy job evaluation.
     if (typeof job === 'function') this.moduleMap.set(url, job = job());
 
-    if (job !== undefined) return job;
+    if (job != null) {
+      const currentImportAssertionType = importAssertionTypeCache.get(job);
+      if (currentImportAssertionType === import_assertions.type) return job;
+
+      try {
+        // To avoid race conditions, wait for previous module to fulfill first.
+        await job.modulePromise;
+      } catch {
+        // If the other job failed with a different `type` assertion, we got
+        // another chance.
+        job = undefined;
+      }
+
+      if (job !== undefined) {
+        const finalFormat = finalFormatCache.get(job);
+        if (
+          import_assertions.type == null ||
+              (import_assertions.type === 'json' && finalFormat === 'json')
```

Comment:
shouldn't this always check if the finalFormat is 'json' and not skip it if it is nullish?

Reply:
If `import_assertions.type` is nullish (I.e. assertionless import) and the previous job succeeded, we return the cached module. Because there was no assertion made, there’s no reason to check what format the module was loaded with. If/when we want to force assertion for some format, we would need to add some additional logic, but that’s not what this PR aims to do. 

> shouldn't this always check if the finalFormat is 'json'

we still want non-JSON modules to load (such as `module`, `commonjs`, etc.), I’m not sure I understand what you mean by that.

Reply:
If `import_assertions.type` is nullish (I.e. assertionless import) and the previous job succeeded, we return the cached module. Because there was no assertion made, there’s no reason to check what format the module was loaded with. If/when we want to force assertion for some format, we would need to add some additional logic, but that’s not what this PR aims to do. 

> shouldn't this always check if the finalFormat is 'json'

we still want non-JSON modules to load (such as `module`, `commonjs`, etc.), I’m not sure I understand what you mean by that.

Reply:
Oh I think I get what you mean, indeed because `json` is the only valid value for an assertion type for now, indeed `json` is the only format acceptable here. I think we should keep the code ready for more values to be added, so when new assertion types are introduced, it won’t have to refactor the whole implementation.

Reply:
the problem is that we have 2 cache entries coalescing to the same module job and that makes me trying to figure out how a userland loader could recreate that behavior very complicated and I am definitely not confident in that collision being safe even with the mitigation I mentioned and requested above. currently we always have 1-1 fully resolved cache key to module instance. The concern is the many-1 model that we introduced by coalescing.